composition-js/src/__tests__/compose.demandControl.test.ts

```diff
@@ -0,0 +1,552 @@
+  });
+
+  describe('when renamed', () => {
+    it.each([
+      [subgraphWithRenamedCost, subgraphWithRenamedListSize],
+      [subgraphWithRenamedCostFromFederationSpec, subgraphWithRenamedListSizeFromFederationSpec]
+    ])('propagates the renamed @cost and @listSize to the supergraph', (costSubgraph: ServiceDefinition, listSizeSubgraph: ServiceDefinition) => {
+      const result = composeServices([costSubgraph, listSizeSubgraph]);
+      assertCompositionSuccess(result);
+      expect(result.hints).toEqual([]);
+
+      // Ensure the new directive names are specified in the supergraph so we can use them during extraction
+      const links = result.schema.schemaDefinition.appliedDirectivesOf("link");
+      const costLink = links.find((link) => link.arguments().url === "https://specs.apollo.dev/cost/v0.1");
+      expect(costLink?.arguments().as).toBe("renamedCost");
+
+      const listSizeLink = links.find((link) => link.arguments().url === "https://specs.apollo.dev/listSize/v0.1");
+      expect(listSizeLink?.arguments().as).toBe("renamedListSize");
+
+      // Ensure the directives are applied to the expected fields with the new names
+      const costDirectiveApplications = fieldWithCost(result)?.appliedDirectivesOf('renamedCost');
+      expect(costDirectiveApplications?.toString()).toMatchString(`@renamedCost(weight: 5)`);
+
+      const argCostDirectiveApplications = argumentWithCost(result)?.appliedDirectivesOf('renamedCost');
+      expect(argCostDirectiveApplications?.toString()).toMatchString(`@renamedCost(weight: 10)`);
+
+      const enumCostDirectiveApplications = enumWithCost(result)?.appliedDirectivesOf('renamedCost');
+      expect(enumCostDirectiveApplications?.toString()).toMatchString(`@renamedCost(weight: 15)`);
+
+      const listSizeDirectiveApplications = fieldWithListSize(result)?.appliedDirectivesOf('renamedListSize');
+      expect(listSizeDirectiveApplications?.toString()).toMatchString(`@renamedListSize(assumedSize: 2000, requireOneSlicingArgument: false)`);
+    });
+  });
+
+  describe('when used on @shareable fields', () => {
+    it('hints about merged @cost arguments', () => {
+      const subgraphA = {
+        name: 'subgraph-a',
+        typeDefs: asFed2SubgraphDocument(gql`
+          extend schema @link(url: "https://specs.apollo.dev/cost/v0.1", import: ["@cost"])
+  
+          type Query {
+            sharedWithCost: Int @shareable @cost(weight: 5)
+          }
+        `)
+      };
+      const subgraphB = {
+        name: 'subgraph-b',
+        typeDefs: asFed2SubgraphDocument(gql`
+          extend schema @link(url: "https://specs.apollo.dev/cost/v0.1", import: ["@cost"])
+  
+          type Query {
+            sharedWithCost: Int @shareable @cost(weight: 10)
+          }
+        `)
+      };
+  
+      const result = composeServices([subgraphA, subgraphB]);
+      assertCompositionSuccess(result);
+      expect(result.hints).toMatchInlineSnapshot(`
+        Array [
```

Comment:
Wondering if we'll get requests to disable the hint.

Reply:
I'm not sure. I think it's important to let people know that we're aggregating costs on shared fields, as my expectation without any hint would be that the original weights would be preserved. I do not know how common it will be to try setting different costs for different subgraphs with a shared field, but I was thinking we could ship this simpler version and get some customer feedback. If customers want that use case, then we could have a future version of the cost spec use joins instead.

I tried implementing this with joins originally, but I ran into a few places where the join spec wasn't sufficient (some issues with renamed directives, not being available on certain locations, and not being fully implemented on the rust side of things).